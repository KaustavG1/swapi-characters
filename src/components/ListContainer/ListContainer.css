.list-container {
  outline: 1px solid blueviolet;
  margin: 20px;
  display: flex;
  flex-wrap: wrap;
  gap: 10px;
  justify-content: center;
}
