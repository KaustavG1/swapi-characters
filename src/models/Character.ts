export interface Character {
  name: string;
  uid: string;
  url: string;
  hair_color: string;
  eye_color: string;
  gender: string;
  homeworld: string;
}
