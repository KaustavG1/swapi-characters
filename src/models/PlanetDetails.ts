export interface PlanetDetails {
  message: string;
  result: {
    properties: {
      diameter: string;
      rotation_period: string;
      orbital_period: string;
      gravity: string;
      population: string;
      climate: string;
      terrain: string;
      surface_water: string;
      created: string;
      edited: string;
      name: string;
      url: string;
    };
    description: string;
    uid: string;
  };
}
