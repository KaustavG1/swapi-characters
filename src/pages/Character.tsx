import { useParams } from "react-router";

function Character() {
  let { id } = useParams();
  console.log(id)

  return (
    <div>Character</div>
  )
}

export default Character
