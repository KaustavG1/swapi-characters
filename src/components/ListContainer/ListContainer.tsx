import { useNavigate } from "react-router";
import Card from "../common/Card/Card";
import EmptyRecords from "../common/EmptyRecords/EmptyRecords";
import { Character } from "../../models/Character";
import "./ListContainer.css";

export interface ListContainerProps {
  data: Character[];
  className?: string;
}

function ListContainer({ data, className }: ListContainerProps) {
  const navigate = useNavigate();

  const handleCardClick = (url: string) => {
    const splitUrl = url?.split("/");
    const id = splitUrl[splitUrl.length - 1];
    navigate(`/${id}`);
  };

  if (data === null) {
    return <EmptyRecords />;
  }

  return (
    <div className={`list-container ${className}`}>
      {data?.map((character) => (
        <Card
          key={character?.url}
          name={character?.name}
          gender={character?.gender}
          homeworld={character?.homeworld}
          onClick={() => handleCardClick(character?.url)}
        />
      ))}
    </div>
  );
}

export default ListContainer;
