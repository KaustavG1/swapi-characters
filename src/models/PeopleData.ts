import { Character } from "./Character";

export interface PeopleData {
  message: string;
  next: string | null;
  previous: string | null;
  results: Character[];
  total_pages: number;
  total_records: number;
}
