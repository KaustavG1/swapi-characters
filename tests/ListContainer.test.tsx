import axios from "axios";
import { it, expect, vi, MockedFunction } from "vitest";
import { render, screen, waitFor } from "@testing-library/react";
import { BrowserRouter } from "react-router";
import { mockPaginationInput } from "./__mocks__/mockPaginationInput";
import ListContainer from "../src/components/ListContainer/ListContainer";
import { Character } from "../src/models/Character";

vi.mock("axios");
(axios.get as MockedFunction<typeof axios>).mockResolvedValue({
  data: mockPaginationInput,
});

afterEach(vi.clearAllMocks);

describe("ListContainer component", () => {
  it("renders 'ListContainer' component success case", async () => {
    const data = mockPaginationInput.map(
      (character: any): Character => ({
        name: character?.name,
        uid: character?.url,
        url: character?.url,
        hair_color: character?.hair_color,
        eye_color: character?.eye_color,
        gender: character?.gender,
        homeworld: "Test Planet",
      })
    );
    const { container } = render(
      <BrowserRouter>
        <ListContainer data={data} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();

    await waitFor(() => {
      expect(screen.getAllByText(/Luke Skywalker/i)[0]).toBeInTheDocument();
    });
  });

  it("renders 'ListContainer' component failure case", async () => {
    (axios.get as MockedFunction<typeof axios>).mockRejectedValue(
      new Error("API Error")
    );
    const { container } = render(
      <BrowserRouter>
        <ListContainer data={[]} />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();

    await waitFor(() => {
      expect(screen.getAllByText(/No Records found./i)[0]).toBeInTheDocument();
    });
  });
});
