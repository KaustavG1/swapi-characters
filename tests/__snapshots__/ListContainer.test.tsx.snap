// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dashboard component > renders 'Dashboard' component failure case 1`] = `
<div>
  <div
    class="blank-page"
  >
    <div
      class="loader"
    >
      Loading ...
    </div>
  </div>
</div>
`;

exports[`Dashboard component > renders 'Dashboard' component success case 1`] = `
<div>
  <div
    class="blank-page"
  >
    <div
      class="loader"
    >
      Loading ...
    </div>
  </div>
</div>
`;

exports[`ListContainer component > renders 'ListContainer' component failure case 1`] = `
<div>
  <div
    class="blank-page"
  >
    <div
      class="empty-record-message"
    >
      No Records found.
    </div>
  </div>
</div>
`;

exports[`ListContainer component > renders 'ListContainer' component success case 1`] = `
<div>
  <div
    class="list-container undefined"
  >
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        Luke Skywalker
      </div>
      <div
        class="card-details"
      >
        Gender: 
        male
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        C-3PO
      </div>
      <div
        class="card-details"
      >
        Gender: 
        n/a
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        R2-D2
      </div>
      <div
        class="card-details"
      >
        Gender: 
        n/a
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        Darth Vader
      </div>
      <div
        class="card-details"
      >
        Gender: 
        male
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        Leia Organa
      </div>
      <div
        class="card-details"
      >
        Gender: 
        female
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        Owen Lars
      </div>
      <div
        class="card-details"
      >
        Gender: 
        male
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        Beru Whitesun lars
      </div>
      <div
        class="card-details"
      >
        Gender: 
        female
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        R5-D4
      </div>
      <div
        class="card-details"
      >
        Gender: 
        n/a
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        Biggs Darklighter
      </div>
      <div
        class="card-details"
      >
        Gender: 
        male
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        Obi-Wan Kenobi
      </div>
      <div
        class="card-details"
      >
        Gender: 
        male
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        Anakin Skywalker
      </div>
      <div
        class="card-details"
      >
        Gender: 
        male
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
    <div
      class="character-card"
    >
      <div
        class="card-name"
      >
        Wilhuff Tarkin
      </div>
      <div
        class="card-details"
      >
        Gender: 
        male
      </div>
      <div
        class="card-details"
      >
        Home Planet: 
        Test Planet
      </div>
    </div>
  </div>
</div>
`;
